filters/diag/diag.go

```diff
@@ -0,0 +1,344 @@
+	if len(args) != 1 {
+		return nil, filters.ErrInvalidFilterParameters
+	}
+
+	if l, ok := args[0].(float64); ok {
+		return &random{int(l)}, nil
+	} else {
+		return nil, filters.ErrInvalidFilterParameters
+	}
+}
+
+func (r *random) ServeHTTP(w http.ResponseWriter, _ *http.Request) {
+	for n := 0; n < r.len; {
+		l := defaultChunkSize
+		if n+l > r.len {
+			l = r.len - n
+		}
+
+		b := make([]byte, l)
+		for i := 0; i < l; i++ {
+			b[i] = randomChars[rand.Intn(len(randomChars))]
+		}
+
+		ni, err := w.Write(b)
+		if err != nil {
+			log.Error(err)
+			return
+		}
+
+		n += ni
+	}
+}
+
+func (r *random) Request(ctx filters.FilterContext) {
+	serve.ServeHTTP(ctx, r)
+}
+
+func (r *random) Response(ctx filters.FilterContext) {}
+
+func (t *throttle) Name() string {
+	switch t.typ {
+	case latency:
+		return LatencyName
+	case bandwidth:
+		return BandwidthName
+	case chunks:
+		return ChunksName
+	case backendLatency:
+		return BackendLatencyName
+	case backendBandwidth:
+		return BackendBandwidthName
+	case backendChunks:
+		return BackendChunksName
+	default:
+		panic("invalid throttle type")
+	}
+}
+
+func parseLatencyArgs(args []interface{}) (int, time.Duration, error) {
+	if len(args) != 1 {
+		return 0, 0, filters.ErrInvalidFilterParameters
+	}
+
+	if msec, ok := args[0].(float64); ok && msec >= 0 {
+		return 0, time.Duration(msec) * time.Millisecond, nil
+	} else {
+		return 0, 0, filters.ErrInvalidFilterParameters
+	}
+}
+
+func parseBandwidthArgs(args []interface{}) (int, time.Duration, error) {
+	if len(args) != 1 {
+		return 0, 0, filters.ErrInvalidFilterParameters
+	}
+
+	kbps, ok := args[0].(float64)
+	if !ok || kbps <= 0 {
+		return 0, 0, filters.ErrInvalidFilterParameters
+	}
+
+	bpms := kbps2bpms(kbps)
+	return defaultChunkSize, time.Duration(float64(defaultChunkSize)/bpms) * time.Millisecond, nil
+}
+
+func parseChunksArgs(args []interface{}) (int, time.Duration, error) {
+	if len(args) != 2 {
+		return 0, 0, filters.ErrInvalidFilterParameters
+	}
+
+	size, ok := args[0].(float64)
+	if !ok || size <= 0 {
+		return 0, 0, filters.ErrInvalidFilterParameters
+	}
+
+	msec, ok := args[1].(float64)
+	if !ok || msec < 0 {
+		return 0, 0, filters.ErrInvalidFilterParameters
+	}
+
+	return int(size), time.Duration(msec) * time.Millisecond, nil
+}
+
+func (t *throttle) CreateFilter(args []interface{}) (filters.Filter, error) {
+	var (
+		chunkSize int
+		delay     time.Duration
+		err       error
+	)
+
+	switch t.typ {
+	case latency, backendLatency:
+		chunkSize, delay, err = parseLatencyArgs(args)
+	case bandwidth, backendBandwidth:
+		chunkSize, delay, err = parseBandwidthArgs(args)
+	case chunks, backendChunks:
+		chunkSize, delay, err = parseChunksArgs(args)
+	default:
+		panic("invalid throttle type")
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &throttle{t.typ, chunkSize, delay}, nil
+}
+
+func (t *throttle) goThrottle(in io.ReadCloser, close bool) io.ReadCloser {
+	if t.chunkSize <= 0 {
+		if t.delay > 0 {
```

Comment:
this is not needed: "A negative or zero duration causes Sleep to return immediately."